Sources/ClientRuntime/Interceptor/Interceptors.swift

```diff
@@ -0,0 +1,448 @@
+    }
+
+    /// - Parameter interceptorFn: The closure to use as the Interceptor hook.
+    public mutating func addReadAfterDeserialization(
+        _ interceptorFn: @escaping (
+            any AfterDeserialization<InputType, OutputType, RequestType, ResponseType, AttributesType>
+        )
+            async throws -> Void
+    ) {
+        self.interceptors.append(AnyInterceptor(readAfterDeserialization: interceptorFn))
+    }
+
+    /// - Parameter interceptorFn: The closure to use as the Interceptor hook.
+    public mutating func addModifyBeforeAttemptCompletion(
+        _ interceptorFn: @escaping (
+            any MutableOutputAfterAttempt<InputType, OutputType, RequestType, ResponseType, AttributesType>
+        )
+            async throws -> Void
+    ) {
+        self.interceptors.append(AnyInterceptor(modifyBeforeAttemptCompletion: interceptorFn))
+    }
+
+    /// - Parameter interceptorFn: The closure to use as the Interceptor hook.
+    public mutating func addReadAfterAttempt(
+        _ interceptorFn: @escaping (any AfterAttempt<InputType, OutputType, RequestType, ResponseType, AttributesType>)
+            async throws -> Void
+    ) {
+        self.interceptors.append(AnyInterceptor(readAfterAttempt: interceptorFn))
+    }
+
+    /// - Parameter interceptorFn: The closure to use as the Interceptor hook.
+    public mutating func addModifyBeforeCompletion(
+        _ interceptorFn: @escaping (
+            any MutableOutputFinalization<InputType, OutputType, RequestType, ResponseType, AttributesType>
+        ) async throws -> Void
+    ) {
+        self.interceptors.append(AnyInterceptor(modifyBeforeCompletion: interceptorFn))
+    }
+
+    /// - Parameter interceptorFn: The closure to use as the Interceptor hook.
+    public mutating func addReadAfterExecution(
+        _ interceptorFn: @escaping (any Finalization<InputType, OutputType, RequestType, ResponseType, AttributesType>)
+            async throws -> Void
+    ) {
+        self.interceptors.append(AnyInterceptor(readAfterExecution: interceptorFn))
+    }
+}
+
+/// This extension is for the convenience of Orchestrator to call the same hook on all interceptors.
+/// Note that it doesn't conform to Interceptor because it is backed by AnyInterceptor, which also
+/// doesn't conform to Interceptor.
+///
+/// These methods throw the last error that occurred when executing a given hook, so if multiple
+/// interceptors fail, only the last error is thrown. The rest are logged (TODO)
+extension Interceptors {
+    internal func readBeforeExecution(context: InterceptorType.InterceptorContextType) async throws {
+        var error: Error?
+        for i in interceptors {
+            do {
+                try await i.readBeforeExecution(context: context)
+            } catch let e {
+                error = e
+            }
+        }
+        if let error = error {
+            throw error
+        }
```

Comment:
This is something I didn't catch, but is there a reason why we don't fail on the first error encountered?

Reply:
I think its a safer contract than failing on the first error encountered, in case you have interceptors dependent on something occurring in a previous interceptor, like setting up some resources. There's still issues, but at least you know your interceptor will run.